text/0000-or-patterns.md

```diff
@@ -0,0 +1,666 @@
+       None => None,
+   };
+   ```
+
+   In `src/libsyntax/parse/token.rs`:
+
+   ```rust
+   BinOp(Minus | Star | Or | And) | OrOr => true,
+   ```
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Simply put, `$p | $q` where `$p` and `$q` are some patterns is now itself
+a legal pattern.
+
+This means that you may for example write:
+
+```rust
+enum Foo<T> {
+    Bar,
+    Baz,
+    Quux(T),
+}
+
+fn main() {
+    match Some(Foo::Bar) {
+        Some(Foo::Bar | Foo::Baz) => { .. },
+        _ => { .. },
+    }
+}
+```
+
+Because `$p | $q` is itself a pattern, this means that you can nest arbitrarily:
+
+```rust
+fn main() {
+    match Some(Foo::Bar) {
+        Some(Foo::Bar | Foo::Quux(0 | 1 | 3)) => { .. },
+        _ => { .. }
+    }
+}
+```
+
+Note that the operator `|` has a low precedence. This means that you
+have to write `foo @ (1 | 2 | 3)` instead of writing `foo @ 1 | 2 | 3`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+To the pattern grammar we introduce the following alternative:
+
+```rust
+pat : pat '|' pat
+    | ...
+    ;
+```
+
+Here `|` has the lowest precedence.
+In particular, the operator `@` binds more tightly than `|` does.
+Thus, `i @ p | q` associates as `i @ (p | q)` as opposed to `(i @ p) | q`.
+
+We then introduce a production:
+
+```rust
+top_pat : '|'? pat ;
+```
+
+We then change the grammar of `let` statements to (as compared to [RFC 2175]):
+
+```rust
+let : LET top_pat maybe_ty_ascription maybe_init_expr ';' ;
+```
+
+We change the grammar of `if let` expressions to:
+
+```rust
+expr_if_let : IF LET top_pat '=' expr_nostruct block (ELSE block_or_if)? ;
+```
+
+And for `while let` expressions:
+
+```rust
+expr_while_let : maybe_label WHILE LET top_pat '=' expr_nostruct block ;
+```
+
+For `for` loop expressions we now have:
+
+```rust
+expr_for : maybe_label FOR top_pat IN expr_nostruct block ;
+```
+
+For `match` expressions we now have:
+
+```rust
+expr_match : MATCH expr_nostruct '{' match_clause* nonblock_match_clause? '}' ;
+match_clause : nonblock_match_clause ',' | block_match_clause ','? ;
+nonblock_match_clause : match_arm (nonblock_expr | block_expr_dot) ;
+block_match_clause : match_arm (block | block_expr) ;
+
+match_arm : maybe_outer_attrs top_pat (IF expr_nostruct)? FAT_ARROW ;
+```
+
+In other words, in all of the contexts where a pattern is currently accepted,
+the compiler will now accept pattern alternations of form `p | q` where
+`p` and `q` are arbitrary patterns.
+
+## Static semantics
+
+1. Given a pattern `p | q` at some depth for some arbitrary patterns `p` and `q`,
+   the pattern is considered ill-formed if:
+
+   + the type inferred for `p` does not unify with the type inferred for `q`, or
+   + the same set of bindings are not introduced in `p` and `q`, or
+   + the type of any two bindings with the same name in `p` and `q` do not unify.
```

Comment:
Or the binding mode; for instance, if you bind the same name as `mut` in `p` and non-`mut` in `q`.